File: src/main/scala/alice/tuprolog/core/data/socket/Datagram_Socket.scala
package alice.tuprolog.core.data.socket

import java.net.{DatagramSocket, InetAddress}
import java.{util => ju}

import alice.tuprolog.core.data.{Term, Var}

@SerialVersionUID(1L)
class Datagram_Socket(theSocket: DatagramSocket) extends AbstractSocket {
  private val _socket: DatagramSocket = theSocket

  override def isClientSocket: Boolean = false

  override def isServerSocket: Boolean = false

  override def isDatagramSocket: Boolean = true

  override def getSocket: DatagramSocket = _socket

  override def getAddress: InetAddress = if (_socket.isBound) _socket.getInetAddress else null

  override def unify(varsUnifiedArg1: ju.List[Var], varsUnifiedArg2: ju.List[Var], theTerm: Term): Boolean = {
    val t = theTerm.getTerm
    t match {
      case variable : Var => variable.unify(varsUnifiedArg1, varsUnifiedArg2, this)

      case socket: AbstractSocket =>
        val addr: InetAddress = socket.getAddress
        socket.isDatagramSocket &&  _socket.getInetAddress.toString == addr.toString
      case _ =>  false
    }

  }

  override def toString: String = _socket.toString
}

File: tuProlog 4.0.0 - SNAPSHOT/src/alice/tuprolog/Datagram_Socket.java
package alice.tuprolog;

import java.net.DatagramSocket;
import java.net.InetAddress;
import java.util.List;

public class Datagram_Socket extends AbstractSocket {
	
	private static final long serialVersionUID = 1L;

	private DatagramSocket socket;

	public Datagram_Socket(DatagramSocket socket) {
		super();
		this.socket = socket;
	}

	@Override
	public boolean isClientSocket() {
		return false;
	}

	@Override
	public boolean isServerSocket() {
		return false;
	}

	@Override
	public boolean isDatagramSocket() {
		return true;
	}

	@Override
	public DatagramSocket getSocket() {
		return socket;
	}

	@Override
	public InetAddress getAddress() {
		if(socket.isBound())return socket.getInetAddress();
		else return null;
	}

	@Override
	boolean unify(List<Var> varsUnifiedArg1, List<Var> varsUnifiedArg2, Term t) {
		t = t.getTerm();
        if (t instanceof Var) {
            return t.unify(varsUnifiedArg1, varsUnifiedArg2, this);
        } else if (t instanceof AbstractSocket && ((AbstractSocket) t).isDatagramSocket()) {
        	InetAddress addr= ((AbstractSocket) t).getAddress();
            return socket.getInetAddress().toString().equals(addr.toString());
        } else {
            return false;
        }
	}
	
	@Override
	public String toString(){
		return socket.toString();
	}
}